arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+#endif
+
+/* CAN message TX bufer setter/getter functions */
+
+static uint8_t mpfs_can_send_message_ready(mpfs_can_instance_t *priv);
+static uint8_t mpfs_can_send_message(mpfs_can_instance_t *priv);
+static uint8_t mpfs_can_send_message_abort(mpfs_can_instance_t *priv);
+static uint32_t mpfs_can_get_tx_buffer_status(mpfs_can_instance_t *priv);
+
+static uint32_t mpfs_can_get_tx_error_count(mpfs_can_instance_t *priv);
+
+#ifdef CONFIG_DEBUG_CAN_INFO
+static bool mpfs_can_get_tx_gte96(mpfs_can_instance_t *priv);
+#endif
+
+/* Driver interface to Nuttx network callbacks */
+
+static int mpfs_ifup(struct net_driver_s *dev);
+static int mpfs_ifdown(struct net_driver_s *dev);
+#ifdef CONFIG_NETDEV_CAN_BITRATE_IOCTL
+static int mpfs_ioctl(struct net_driver_s *dev, int cmd, unsigned long arg);
+#endif
+
+/****************************************************************************
+ * Private Function
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_can_retrieve_rx_frame
+ *
+ * Description:
+ *  Retrieve CAN 2.0B frame from RX Buffer
+ *
+ * Input Parameters:
+ *  priv    - Pointer to the private CAN driver state structure
+ *  cf      - Pointer to CAN frame structure
+ *
+ * Returned Value:
+ *  This function returns CAN_OK on successful retrieval of CAN frame else
+ *  it returns CAN_ERR
+ *
+ * Assumptions:
+ *  Frame format word is already parsed in advance and provided as 'ffw' arg
+ *
+ ****************************************************************************/
+
+static bool mpfs_can_retrieve_rx_frame(mpfs_can_instance_t *priv,
+                                       struct can_frame *cf)
+{
+  uint8_t dlc;
+  bool ide;
+  bool rtr;
+  uint8_t i;
+  pmpfs_can_msgobject_t pmsg = priv->rx_msg;
+
+  /* CAN ID & EFF & RTR Flags */
+
+  ide = (bool)(pmsg->msg_ctrl & MPFS_CAN_RX_MSG_CTRL_CMD_IDE);
+  rtr = (bool)(pmsg->msg_ctrl & MPFS_CAN_RX_MSG_CTRL_CMD_RTR);
+  cf->can_id = mpfs_can_msgid_to_canid(pmsg->id, ide, rtr);
+
+  /* Filter check */
+
+#ifdef CONFIG_NETDEV_CAN_FILTER_IOCTL
+  uint32_t can_id_masked = cf->can_id & CAN_ERR_MASK;
+
+  if (expect_true(!priv->filter.use_mask_filter
+      && !priv->filter.use_range_filter)
+      || (priv->filter.use_mask_filter
+      && (can_id_masked & priv->filter.mask) == priv->filter.code)
+      || (priv->filter.use_range_filter
+      && can_id_masked >= priv->filter.range_lower
+      && can_id_masked <= priv->filter.range_upper))
+  {
+    goto accepted;
+  }
+
+  return CAN_ERR;
+
+accepted:
+#endif  /* CONFIG_NETDEV_CAN_FILTER_IOCTL */
+
+  /* DLC */
+
+  dlc = (pmsg->msg_ctrl & MPFS_CAN_RX_MSG_CTRL_CMD_DLC)
+         >> MPFS_CAN_RX_MSG_CTRL_CMD_DLC_SHIFT;
+  if (dlc <= 8)
+    {
+     cf->can_dlc = dlc;
+    }
+  else
+    {
+      canerr("DLC = %d is out of range\n", dlc);
+      return CAN_ERR;
+    }
+
+  /* Data (big endian) */
+
+  for (i = 0; i < dlc; i++)
```

Comment:
This is just unnecessarily complicated. The cf->data is always 32-bit aligned in nuttx can frames (intentionally). Also your pmsg->data_high and low are 32-bit aligned.

So you can just:
          *(uint32_t *)&cf->data[0] = __builtin_bswap32(pmsg->data_high);
          *(uint32_t *)&cf->data[4] = __builtin_bswap32(pmsg->data_low);

If in doubt, add a debugassertion for the alignment.....

Reply:
thanks for the optimized suggestion!

Reply:
yes, this's much more elegant operation. i forgot about that. thanks!